Stratis.Bitcoin/BlockStore/BlockRepository.cs

```diff
@@ -139,58 +141,93 @@ public Task<uint256> GetTrxBlockIdAsync(uint256 trxid)
                     this.PerformanceCounter.AddRepositoryHitCount(1);
                     return blockid.Value;
                 }
-			});
-		}		
+            });
+        }
 
-		public uint256 BlockHash { get; private set; }
-		public bool TxIndex { get; private set; }
+        public uint256 BlockHash { get; private set; }
+        public bool TxIndex { get; private set; }
 
-		public Task PutAsync(uint256 nextBlockHash, List<Block> blocks)
-		{
-			Guard.NotNull(nextBlockHash, nameof(nextBlockHash));
-			Guard.NotNull(blocks, nameof(blocks));
+        virtual protected void OnInsertBlocks(List<Block> blocks)
+        {
+            var transactions = new List<(Transaction, Block)>();
 
-			// dbreeze is faster if sort ascending by key in memory before insert
-			// however we need to find how byte arrays are sorted in dbreeze this link can help 
-			// https://docs.google.com/document/pub?id=1IFkXoX3Tc2zHNAQN9EmGSXZGbabMrWmpmVxFsLxLsw
+            var byteListComparer = new ByteListComparer();
 
-			return this.session.Do(() =>
-			{
-				foreach (var block in blocks)
-				{
-					var blockId = block.GetHash();
-
-					// if the block is already in store don't write it again
-					var item = this.session.Transaction.Select<byte[], Block>("Block", blockId.ToBytes());
-                    if (!item.Exists)
-                    {
-                        this.PerformanceCounter.AddRepositoryMissCount(1);
-                        this.PerformanceCounter.AddRepositoryInsertCount(1);
-                        this.session.Transaction.Insert<byte[], Block>("Block", blockId.ToBytes(), block);
-
-                        if (this.TxIndex)
-                        {
-                            // index transactions
-                            foreach (var transaction in block.Transactions)
-                            {
-                                var trxId = transaction.GetHash();
-                                this.PerformanceCounter.AddRepositoryInsertCount(1);
-                                this.session.Transaction.Insert<byte[], uint256>("Transaction", trxId.ToBytes(), blockId);
-                            }
-                        }
-                    }
-                    else
-                    {
-                        this.PerformanceCounter.AddRepositoryHitCount(1);
-                    }
+            var blockDict = new Dictionary<uint256, Block>();
```

Comment:
Dictionary<byte[], Block>() will avoid the double ToByte() work (sort and serialize).